Added NETCONF notification subscribe and processing capability

nhattrinh opened the pull request:
Added a NetconfNotification class which subclasses a Thread class to listen for notifications from the device as a result of establishing a subscription with a device. Added notify_wait into the Netconf class to trigger event, and added subscribe function in Netconf class to create the NetconfNotification thread, start listening for notifications and add the thread to active notifications.

At the start of the subscribe action, terminal will show the banner:
![image](https://user-images.githubusercontent.com/9090035/102155465-7196e980-3e30-11eb-8d69-50b009f60cab.png)

When listening for notifications, terminal will show the following:
![image](https://user-images.githubusercontent.com/9090035/102155367-490eef80-3e30-11eb-8018-43ec61abe2c9.png)

When the subscribe action times out, the terminal will show:
![image](https://user-images.githubusercontent.com/9090035/102155513-896e6d80-3e30-11eb-8530-17530e61b9d4.png)



Sample log: [job_output.log](https://github.com/CiscoTestAutomation/yang/files/5692559/job_output.log)

LenzDong commented on connector/src/yang/connector/netconf.py:
```diff
@@ -1,11 +1,17 @@
 
 import re
+import time
+from datetime import datetime as dt
```
please remove duplicate import, line 12 already import `datetime`

LenzDong commented on connector/src/yang/connector/netconf.py:
```diff
@@ -1,11 +1,17 @@
+import time
+from datetime import datetime as dt
+import atexit
+import logging
```
please remove this duplicate import

ahersi2-cisco commented:
@yuekyang can you please take a look at this PR as a final check? Thank you!

yuekyang commented on connector/src/yang/connector/netconf.py:
```diff
@@ -849,3 +890,64 @@ def _request(self, msg):
+            td = t2 - t1
+            notif = self.device.take_notification(timeout=1)
+
+            logger.info('Listening for notifications from subscribe stream, {} seconds elapsed'.format(td.seconds))
```
This line is better to stay with `t2 = dt.now()`.

yuekyang commented on connector/src/yang/connector/netconf.py:
```diff
@@ -849,3 +890,64 @@ def _request(self, msg):
                                           'for an rpc-reply.')
+
+
+class NetconfNotification(Thread):
```
How about calling it `NotificationThread` ?

yuekyang commented on connector/src/yang/connector/netconf.py:
```diff
@@ -210,6 +216,18 @@ def __init__(self, *args, **kwargs):
 
+    active_notifications = {}
+
+    def subscribe(self, request, steps):
```
`steps` is not used here. It can be removed.

yuekyang commented on connector/src/yang/connector/netconf.py:
```diff
@@ -210,6 +216,18 @@ def __init__(self, *args, **kwargs):
             self, session=session, device_handler=device_handler,
             timeout=self.timeout)
 
+    active_notifications = {}
```
Do you mean:
`self.active_notifications = {}`

yuekyang commented on connector/src/yang/connector/netconf.py:
```diff
@@ -210,6 +216,18 @@ def __init__(self, *args, **kwargs):
+
+        subscribe_thread.start()
+
+        self.active_notifications[self] = subscribe_thread
```
I'm not convinced to use `self` as a key of `self.active_notifications`. For one Netconf session, how many `subscribe_thread` are allowed?

nhattrinh replied:
should I use a unique ID for the key of `self.active_notifications`? Currently for one NETCONF session, there is no limit on how many `subscribe_thread` are allowed.

nhattrinh replied:
I used a unique id to reference the notification thread

yuekyang commented on connector/src/yang/connector/netconf.py:
```diff
@@ -414,6 +432,29 @@ def disconnect(self):
 
         self.session.close()
 
+    def notify_wait(self, steps):
```
`steps` is a concept in pyATS test cases. Is it possible to avoid passing in `steps`?

nhattrinh replied:
No, I haven't seen another alternative from my current standpoint.

yuekyang reviewed (commented):
Thanks for the `class NetconfNotification`. I think it's useful.

nhattrinh commented:
I have addressed all the comments, please review my changes.

nhattrinh commented:
Closed this PR and created a new PR (https://github.com/CiscoTestAutomation/yang/pull/28).

The new PR has addressed all the review comments in this PR.
